File: Services/PhishNet/PhishNetTourService.swift
//
//  PhishNetTourService.swift
//  PhishQS
//
//  Created by Claude on 9/4/25.
//

import Foundation

/// Service for fetching tour-related data from Phish.net API
/// Provides authoritative tour organization, show counts, and tour positions using Phish.net data
class PhishNetTourService {
    
    private let phishNetClient: PhishAPIClient
    
    init(phishNetClient: PhishAPIClient = PhishAPIClient.shared) {
        self.phishNetClient = phishNetClient
    }
    
    // MARK: - Tour Show Methods
    
    /// Fetch all shows for a specific tour by year and tour name
    /// Provides complete tour show listing from Phish.net API
    func fetchTourShows(year: String, tourName: String) async throws -> [Show] {
        let allYearShows = try await phishNetClient.fetchShows(forYear: year)
        
        // Filter shows by tour name (exact match)
        let tourShows = allYearShows.filter { show in
            show.tour_name == tourName
        }
        
        // Sort by date to ensure proper order
        return tourShows.sorted { $0.showdate < $1.showdate }
    }
    
    /// Get all shows for a year (used for tour detection)
    func fetchAllShowsForYear(_ year: String) async throws -> [Show] {
        return try await phishNetClient.fetchShows(forYear: year)
    }
    
    // MARK: - Tour Position Calculation
    
    /// Calculate tour position for a specific show
    /// Determines show number within tour using Phish.net show ordering
    func calculateTourPosition(for showDate: String, tourName: String) async throws -> TourShowPosition? {
        // Extract year from show date
        let year = String(showDate.prefix(4))
        
        // Get all shows for the tour
        let tourShows = try await fetchTourShows(year: year, tourName: tourName)
        
        // Find the show in the tour
        guard let currentShowIndex = tourShows.firstIndex(where: { $0.showdate == showDate }) else {
            return nil
        }
        
        return TourShowPosition(
            tourName: tourName,
            showNumber: currentShowIndex + 1,
            totalShows: tourShows.count,
            tourYear: year
        )
    }
    
    // MARK: - Tour Detection
    
    /// Extract tour name from a show (if available)
    /// Uses Phish.net tour_name field for consistent tour identification
    func extractTourFromShow(_ show: Show) -> String? {
        return show.tour_name
    }
    
    /// Get tour name for a specific show date by fetching show data
    func getTourNameForShow(date: String) async throws -> String? {
        let year = String(date.prefix(4))
        let allYearShows = try await fetchAllShowsForYear(year)
        
        guard let show = allYearShows.first(where: { $0.showdate == date }) else {
            return nil
        }
        
        return extractTourFromShow(show)
    }
    
    // MARK: - Venue Run Calculation
    
    /// Calculate venue runs for a set of shows
    /// Uses Phish.net show data to determine multi-night venue runs
    /// Matches the server-side venue run calculation logic for consistency
    func calculateVenueRuns(for shows: [Show]) -> [String: VenueRun] {
        print("🏟️  PhishNetTourService.calculateVenueRuns for \(shows.count) shows")
        var venueRuns: [String: VenueRun] = [:]
        
        // Sort shows by date to ensure proper chronological order
        let sortedShows = shows.sorted { $0.showdate < $1.showdate }
        
        // Log show venues for debugging
        for show in sortedShows.prefix(3) {
            print("   📅 \(show.showdate): \(show.venue ?? "no venue"), \(show.city ?? "no city"), \(show.state ?? "no state")")
        }
        if sortedShows.count > 3 {
            print("   ... and \(sortedShows.count - 3) more shows")
        }
        
        // Group consecutive shows by venue
        var venueGroups: [[Show]] = []
        var currentGroup: [Show] = []
        var currentVenue: String = ""
        
        for show in sortedShows {
            let showVenue = show.venue ?? "Unknown Venue"
            
            if showVenue == currentVenue && !currentGroup.isEmpty {
                // Same venue, add to current group
                currentGroup.append(show)
            } else {
                // New venue or first show - finish previous group if it exists
                if !currentGroup.isEmpty {
                    venueGroups.append(currentGroup)
                }
                // Start new group
                currentGroup = [show]
                currentVenue = showVenue
            }
        }
        
        // Don't forget the last group
        if !currentGroup.isEmpty {
            venueGroups.append(currentGroup)
        }
        
        print("   🎪 Found \(venueGroups.count) venue groups")
        
        // Generate venue runs for multi-night runs only
        for group in venueGroups {
            if group.count > 1 {
                print("   🏟️  Multi-night run: \(group[0].venue) (\(group.count) nights)")
                
                // Multi-night run - create venue run objects
                let showDates = group.map { $0.showdate }
                
                for (index, show) in group.enumerated() {
                    let venueRun = VenueRun(
                        venue: show.venue ?? "Unknown Venue",
                        city: show.city ?? "Unknown City",
                        state: show.state,
                        nightNumber: index + 1,
                        totalNights: group.count,
                        showDates: showDates
                    )
                    venueRuns[show.showdate] = venueRun
                    print("     📅 \(show.showdate): N\(index + 1)/\(group.count)")
                }
            } else {
                print("   🏟️  Single night: \(group[0].venue ?? "Unknown Venue") (\(group[0].showdate))")
            }
        }
        
        print("   ✅ Generated \(venueRuns.count) venue run objects")
        return venueRuns
    }
    
    /// Get venue run for a specific show date within a tour
    func getVenueRun(for showDate: String, in tourShows: [Show]) -> VenueRun? {
        let venueRuns = calculateVenueRuns(for: tourShows)
        return venueRuns[showDate]
    }
    
    // MARK: - Convenience Methods
    
    /// Get complete tour context for a show (position + venue run)
    func getTourContext(for showDate: String) async throws -> (tourPosition: TourShowPosition?, venueRun: VenueRun?) {
        print("🎪 PhishNetTourService.getTourContext for \(showDate)")
        
        guard let tourName = try await getTourNameForShow(date: showDate) else {
            print("   ⚠️  No tour name found for \(showDate)")
            return (nil, nil)
        }
        
        print("   🎪 Found tour: \(tourName)")
        
        let year = String(showDate.prefix(4))
        let tourShows = try await fetchTourShows(year: year, tourName: tourName)
        print("   📋 Found \(tourShows.count) tour shows")
        
        let tourPosition = try await calculateTourPosition(for: showDate, tourName: tourName)
        print("   🎯 Tour position: \(tourPosition?.showNumber ?? 0)/\(tourPosition?.totalShows ?? 0)")
        
        let venueRun = getVenueRun(for: showDate, in: tourShows)
        if let venueRun = venueRun {
            print("   🏟️  Venue run: N\(venueRun.nightNumber)/\(venueRun.totalNights) at \(venueRun.venue)")
        } else {
            print("   🏟️  No venue run (single night show)")
        }
        
        return (tourPosition, venueRun)
    }
    
    // MARK: - Tour Name Normalization
    
    /// Normalize tour names for consistency
    /// Handles variations in tour name formatting between API responses
    static func normalizeTourName(_ tourName: String) -> String {
        // Handle common tour name variations
        if tourName == "Early Summer Tour 2025" {
            return "2025 Early Summer Tour"
        }
        if tourName == "Late Summer Tour 2025" {
            return "2025 Late Summer Tour"
        }
        
        // Add more mappings as needed
        return tourName
    }
}

File: Server/Services/PhishNetTourService.js
/**
 * PhishNetTourService.js
 * 
 * Service for fetching tour-related data from Phish.net API
 * Replaces Phish.in as the authoritative source for tour organization, show counts, and venue runs
 * 
 * Created by Claude on 9/4/25
 */

import { PhishNetClient } from '../API/PhishNetClient.js';
import { TourScheduleService } from './TourScheduleService.js';
import LoggingService from './LoggingService.js';

export class PhishNetTourService {
    
    constructor(apiKey) {
        this.phishNetClient = new PhishNetClient(apiKey);
        this.tourScheduleService = new TourScheduleService();
    }
    
    // MARK: - Tour Show Methods
    
    /**
     * Fetch all shows for a specific tour by year and tour name
     * Replaces Phish.in getCachedTourShows functionality
     * 
     * @param {string} year - Year to fetch shows for (e.g., '2025')
     * @param {string} tourName - Tour name to filter by (e.g., '2025 Summer Tour')
     * @returns {Promise<Array>} Array of show objects for the tour
     */
    async fetchTourShows(year, tourName) {
        try {
            const allYearShows = await this.phishNetClient.fetchShows(year);
            
            // Filter shows by tour name (exact match) - use 'tourname' field from Phish.net
            const tourShows = allYearShows.filter(show => show.tourname === tourName);
            
            LoggingService.info(`Found ${tourShows.length} shows for ${tourName}`);
            
            // Sort by date to ensure proper order (though API should already be ordered)
            return tourShows.sort((a, b) => a.showdate.localeCompare(b.showdate));
        } catch (error) {
            console.error(`Error fetching tour shows for ${tourName} in ${year}:`, error);
            return [];
        }
    }
    
    /**
     * Get all shows for a year (used for tour detection)
     * 
     * @param {string} year - Year to fetch shows for
     * @returns {Promise<Array>} Array of all shows for the year
     */
    async fetchAllShowsForYear(year) {
        return await this.phishNetClient.fetchShows(year);
    }
    
    // MARK: - Tour Position Calculation
    
    /**
     * Calculate tour position for a specific show
     * Replaces Phish.in fetchTourPosition functionality
     * 
     * @param {string} showDate - Show date in YYYY-MM-DD format
     * @param {string} tourName - Tour name
     * @returns {Promise<Object|null>} TourShowPosition object or null
     */
    async calculateTourPosition(showDate, tourName) {
        try {
            // First check if we have schedule data for accurate total show counts
            if (this.tourScheduleService.isTourSupported(tourName)) {
                return this.tourScheduleService.calculateTourPosition(showDate, tourName);
            }
            
            // Fallback to setlist-based calculation for unsupported tours
            const year = showDate.substring(0, 4);
            
            // Get all played shows for the tour (from setlist API)
            const tourShows = await this.fetchTourShows(year, tourName);
            
            // Find the show in the tour
            const currentShowIndex = tourShows.findIndex(show => show.showdate === showDate);
            if (currentShowIndex === -1) {
                return null;
            }
            
            // Return position (1-indexed) - note: totalShows will only reflect played shows
            return {
                tourName: tourName,
                showNumber: currentShowIndex + 1,
                totalShows: tourShows.length,
                tourYear: year
            };
        } catch (error) {
            console.error(`Error calculating tour position for ${showDate}:`, error);
            return null;
        }
    }
    
    // MARK: - Tour Detection
    
    /**
     * Extract tour name from a show (if available)
     * Replaces Phish.in tour_name field usage
     * 
     * @param {Object} show - Show object
     * @returns {string|null} Tour name or null
     */
    extractTourFromShow(show) {
        return show.tourname || null;
    }
    
    /**
     * Get tour name for a specific show date by fetching show data
     * 
     * @param {string} showDate - Show date in YYYY-MM-DD format
     * @returns {Promise<string|null>} Tour name or null
     */
    async getTourNameForShow(showDate) {
        try {
            const year = showDate.substring(0, 4);
            const allYearShows = await this.fetchAllShowsForYear(year);
            
            const show = allYearShows.find(show => show.showdate === showDate);
            if (!show) {
                return null;
            }
            
            return this.extractTourFromShow(show);
        } catch (error) {
            console.error(`Error getting tour name for ${showDate}:`, error);
            return null;
        }
    }
    
    // MARK: - Venue Run Calculation
    
    /**
     * Calculate venue runs for a set of shows
     * Replaces Phish.in venue run detection
     * 
     * @param {Array} shows - Array of show objects
     * @returns {Object} Object mapping show dates to VenueRun objects
     */
    calculateVenueRuns(shows) {
        const venueRuns = {};
        
        // Group shows by venue (venue + city + state as key)
        const venueGroups = {};
        
        shows.forEach(show => {
            const venueKey = `${show.venue || 'Unknown'}-${show.city || 'Unknown'}-${show.state || 'Unknown'}`;
            if (!venueGroups[venueKey]) {
                venueGroups[venueKey] = [];
            }
            venueGroups[venueKey].push(show);
        });
        
        // Process each venue group
        Object.values(venueGroups).forEach(venueShows => {
            const sortedShows = venueShows.sort((a, b) => a.showdate.localeCompare(b.showdate));
            
            // Only create venue runs for multi-night stands
            if (sortedShows.length > 1) {
                const showDates = sortedShows.map(show => show.showdate);
                const firstShow = sortedShows[0];
                
                sortedShows.forEach((show, index) => {
                    const venueRun = {
                        venue: firstShow.venue || 'Unknown Venue',
                        city: firstShow.city || 'Unknown City',
                        state: firstShow.state || null,
                        nightNumber: index + 1,
                        totalNights: sortedShows.length,
                        showDates: showDates
                    };
                    
                    venueRuns[show.showdate] = venueRun;
                });
            }
        });
        
        return venueRuns;
    }
    
    /**
     * Get venue run for a specific show date within a tour
     * 
     * @param {string} showDate - Show date in YYYY-MM-DD format
     * @param {Array} tourShows - Array of tour show objects
     * @returns {Object|null} VenueRun object or null
     */
    getVenueRun(showDate, tourShows) {
        const venueRuns = this.calculateVenueRuns(tourShows);
        return venueRuns[showDate] || null;
    }
    
    // MARK: - Convenience Methods
    
    /**
     * Get complete tour context for a show (position + venue run)
     * 
     * @param {string} showDate - Show date in YYYY-MM-DD format
     * @returns {Promise<Object>} Object with tourPosition and venueRun properties
     */
    async getTourContext(showDate) {
        try {
            const tourName = await this.getTourNameForShow(showDate);
            if (!tourName) {
                return { tourPosition: null, venueRun: null };
            }
            
            const year = showDate.substring(0, 4);
            const tourShows = await this.fetchTourShows(year, tourName);
            
            const tourPosition = await this.calculateTourPosition(showDate, tourName);
            const venueRun = this.getVenueRun(showDate, tourShows);
            
            return { tourPosition, venueRun };
        } catch (error) {
            console.error(`Error getting tour context for ${showDate}:`, error);
            return { tourPosition: null, venueRun: null };
        }
    }
    
    // MARK: - Tour Name Normalization
    
    /**
     * Normalize tour names between different API sources
     * Maps Phish.in format to Phish.net format
     * 
     * @param {string} tourName - Tour name to normalize
     * @returns {string} Normalized tour name
     */
    static normalizeTourName(tourName) {
        // Handle common tour name variations
        if (tourName === 'Summer Tour 2025') {
            return '2025 Summer Tour';
        }
        
        // Add more mappings as needed
        return tourName;
    }
    
    /**
     * Get current tour from latest show
     * Used by statistics generation to determine active tour
     * 
     * @returns {Promise<string|null>} Current tour name or null
     */
    async getCurrentTour() {
        try {
            const latestShow = await this.phishNetClient.fetchLatestShow();
            if (!latestShow) {
                return null;
            }
            
            return this.extractTourFromShow(latestShow);
        } catch (error) {
            console.error('Error getting current tour:', error);
            return null;
        }
    }
}